File: chart.js
/* The one chart component (SEB-61 / U6) -- browser half.
 *
 * The chart is already fully drawn in the page's own markup by tools/chart.py
 * at generation time: headline, change line, plot, gutter, badge and date
 * labels all render with no JavaScript and under `prefers-reduced-motion`.
 * This file only adds what genuinely needs a browser: redrawing the plot
 * when a reader picks a different range pill, and answering hover with a
 * point's value and date. Nothing here can appear before the static chart
 * already has.
 *
 * Stdlib DOM only, no build step, matching every other script on this site.
 */
(function () {
  "use strict";
  var WIDTH = 1000, HEIGHT = 200;
  var INDIGO = "#5A55E0", DOWN = "#D0453B", BACKFILL = "#b9b6b4";
  var WINDOWS = { "1w": 7, "1m": 30, "3m": 90, "1y": 365 };

  function fmtPct(v) {
    if (v === null || v === undefined || !isFinite(v)) return "—";
    var mag = Math.abs(v);
    return (v >= 0 ? "+" : "−") + mag.toFixed(mag >= 10 ? 1 : 2) + "%";
  }

  // Month names are reader-facing text, so they come from the copy deck via a
  // data attribute, the same way the change wording does -- never typed here.
  function longDate(iso, months) {
    var parts = iso.split("-").map(Number);
    return parts[2] + " " + months[parts[1] - 1] + " " + parts[0];
  }

  function sliceRange(all, key) {
    if (key === "all" || !WINDOWS[key]) return all;
    var last = new Date(all[all.length - 1].date + "T00:00:00Z");
    var cutoff = new Date(last.getTime() - WINDOWS[key] * 86400000);
    return all.filter(function (p) { return new Date(p.date + "T00:00:00Z") >= cutoff; });
  }

  function runs(pts) {
    var out = [], cur = [];
    for (var i = 0; i < pts.length; i++) {
      var p = pts[i], prev = cur[cur.length - 1];
      if (prev) {
        var days = (new Date(p.date + "T00:00:00Z") - new Date(prev.date + "T00:00:00Z")) / 86400000;
        if (days === 1 && p.backfill === prev.backfill) { cur.push(p); continue; }
        out.push(cur);
        cur = [];
      }
      cur.push(p);
    }
    if (cur.length) out.push(cur);
    return out;
  }

  function render(root, pts, copy) {
    var n = pts.length;
    // Scaled to the data's own range, not anchored to zero -- see chart.py's
    // _geometry for why (a series far from zero would otherwise flatten).
    var values = pts.map(function (p) { return p.index_pct; });
    var lo = Math.min.apply(null, values);
    var hi = Math.max.apply(null, values);
    var pad = (hi - lo) * 0.12 || 1;
    lo -= pad; hi += pad;
    var indexOf = {};
    pts.forEach(function (p, i) { indexOf[p.date] = i; });
    var x = function (i) { return n > 1 ? (i / (n - 1)) * WIDTH : 0; };
    var y = function (v) { return HEIGHT - ((v - lo) / (hi - lo)) * HEIGHT; };

    var svg = "";
    runs(pts).forEach(function (run) {
      var colour = run[0].backfill ? BACKFILL : INDIGO;
      if (run.length === 1) {
        var i0 = indexOf[run[0].date];
        svg += '<circle cx="' + x(i0).toFixed(1) + '" cy="' + y(run[0].index_pct).toFixed(1) +
          '" r="2.5" fill="' + colour + '"/>';
        return;
      }
      var seg = run.map(function (p) {
        return x(indexOf[p.date]).toFixed(1) + " " + y(p.index_pct).toFixed(1);
      }).join(" L ");
      if (!run[0].backfill) {
        var firstI = indexOf[run[0].date], lastI = indexOf[run[run.length - 1].date];
        svg += '<path d="M ' + x(firstI).toFixed(1) + " " + HEIGHT + " L " + seg + " L " +
          x(lastI).toFixed(1) + " " + HEIGHT + ' Z" fill="url(#chartFade)"/>';
      }
      svg += '<path d="M ' + seg + '" fill="none" stroke="' + colour +
        '" stroke-width="1.6" vector-effect="non-scaling-stroke"/>';
    });

    var last = pts[n - 1], first = pts[0];
    var badgeY = y(last.index_pct);
    svg += '<line x1="0" x2="' + WIDTH + '" y1="' + badgeY.toFixed(1) + '" y2="' + badgeY.toFixed(1) +
      '" stroke="' + INDIGO + '" stroke-width="1" stroke-dasharray="1 3" vector-effect="non-scaling-stroke"/>';
    root.querySelector(".chart-svg").innerHTML =
      root.querySelector(".chart-svg defs").outerHTML + svg;

    var badgePct = badgeY / HEIGHT * 100;
    root.querySelector(".chart-badge").style.top = badgePct + "%";
    root.querySelector(".chart-badge").textContent = fmtPct(last.index_pct);

    var gutterHtml = "";
    [[0, hi], [50, (hi + lo) / 2], [100, lo]].forEach(function (pair) {
      if (Math.abs(pair[0] - badgePct) < 12) return;
      gutterHtml += '<span class="chart-gutter-label" style="top:' + pair[0] + '%">' +
        fmtPct(pair[1]) + "</span>";
    });
    root.querySelector(".chart-gutter").innerHTML = gutterHtml;

    root.querySelector(".chart-value").textContent = fmtPct(last.index_pct);
    var changeEl = root.querySelector(".chart-change");
    var changePts = last.index_pct - first.index_pct;
    if (Math.round(changePts * 100) / 100 === 0) {
      changeEl.textContent = copy.changeNone;
      changeEl.style.color = "var(--muted)";
    } else {
      var sign = changePts >= 0 ? "+" : "−";
      var text = sign + Math.abs(changePts).toFixed(2) + " points";
      if (first.index_pct !== 0 && (first.index_pct > 0) === (last.index_pct > 0)) {
        var rel = changePts / Math.abs(first.index_pct) * 100;
        text += ", " + (changePts >= 0 ? copy.changeUp : copy.changeDown) + " " + Math.abs(rel).toFixed(1) + "%";
      }
      changeEl.textContent = text;
      changeEl.style.color = changePts >= 0 ? INDIGO : DOWN;
    }

    var fracs = [0, 1 / 3, 2 / 3, 1];
    var idxs = Array.from(new Set(fracs.map(function (f) { return Math.round(f * (n - 1)); })));
    var datesHtml = "";
    idxs.forEach(function (i, j) {
      var align = j === 0 ? "start" : (j === idxs.length - 1 ? "end" : "middle");
      datesHtml += '<span class="chart-date" style="text-align:' + align + '">' +
        longDate(pts[i].date, copy.months) + "</span>";
    });
    root.querySelector(".chart-dates").innerHTML = datesHtml;
  }

  function wire(root) {
    var all;
    try { all = JSON.parse(root.dataset.points || "[]"); } catch (e) { all = []; }
    if (all.length < 2) return;
    var current = all;
    var copy = {
      changeNone: root.dataset.copyChangeNone,
      changeUp: root.dataset.copyChangeUp,
      changeDown: root.dataset.copyChangeDown,
      months: (root.dataset.copyMonths || "").split("|")
    };

    var pills = root.querySelectorAll(".chart-pill");
    pills.forEach(function (btn) {
      btn.addEventListener("click", function () {
        current = sliceRange(all, btn.dataset.range);
        if (current.length < 2) return;
        render(root, current, copy);
        pills.forEach(function (b) { b.classList.toggle("is-on", b === btn); });
      });
    });

    var wrap = root.querySelector(".chart-plot-wrap");
    var hoverEl = root.querySelector(".chart-hover");
    if (!wrap || !hoverEl) return;
    wrap.addEventListener("mousemove", function (e) {
      var rect = wrap.getBoundingClientRect();
      if (!rect.width || current.length < 2) return;
      var frac = Math.min(1, Math.max(0, (e.clientX - rect.left) / rect.width));
      var i = Math.round(frac * (current.length - 1));
      var p = current[i];
      if (!p) return;
      hoverEl.hidden = false;
      hoverEl.style.left = (current.length > 1 ? (i / (current.length - 1)) * 100 : 0) + "%";
      hoverEl.textContent = fmtPct(p.index_pct) + " — " + longDate(p.date, copy.months);
    });
    wrap.addEventListener("mouseleave", function () { hoverEl.hidden = true; });
  }

  document.documentElement.classList.add("js");
  window.MarginChart = { wire: wire };
})();

File: tools/chart.py
#!/usr/bin/env python3
"""The one chart component (SEB-61 / U6), server-rendered.

Implements the anatomy specified on SEB-61: a headline value, the change over
the window shown, a single indigo line over a fading area, a right-hand gutter
of up to three value labels, a dotted current-value line with a badge, up to
four date labels along the bottom, and range pills offered only where the
data can fill them.

Rendered once here, at generation time in Python, from the same `history`
list that produces the number printed above it -- so the chart can never show
a different "now" than the headline stat on the page (SEB-60/D1). Nothing
here is fabricated: every path, label and pill is a function of the points it
is given.

A companion, `chart.js`, re-runs this same geometry in the browser only to
redraw the plot when a reader picks a different range pill, and to answer
hover with a point's value and date. Nothing it does can appear before this
module's static markup has already painted the chart -- the chart is fully
present with JavaScript off, or `prefers-reduced-motion` on.

Stdlib only.
"""

import datetime as dt

INDIGO = "#5A55E0"
DOWN = "#D0453B"
BACKFILL = "#b9b6b4"
AREA_TOP = "#DCDFFA"

WIDTH = 1000
HEIGHT = 200

# key, label, window in days. 1h/6h/1d need intraday points a one-a-day
# history structurally cannot hold, so they are never offered here -- but the
# check is generic, not hard-coded to "this chart is daily", so the same
# table works if a chart is ever built over hourly points instead.
RANGES = [
    ("1h", "1h", 1 / 24),
    ("6h", "6h", 6 / 24),
    ("1d", "1d", 1),
    ("1w", "1w", 7),
    ("1m", "1m", 30),
    ("3m", "3m", 90),
    ("1y", "1y", 365),
    ("all", "All", None),
]

DAY_NAMES = ["January", "February", "March", "April", "May", "June", "July",
             "August", "September", "October", "November", "December"]


def fmt_pct(v):
    """Mirrors the page's own `pct()` -- same rounding, same sign glyph."""
    if v is None:
        return "—"
    sign = "+" if v >= 0 else "−"
    mag = abs(v)
    return sign + f"{mag:.{1 if mag >= 10 else 2}f}" + "%"


def long_date(iso):
    y, m, d = (int(x) for x in iso.split("-"))
    return f"{d} {DAY_NAMES[m - 1]} {y}"


def _parsed(points):
    out = []
    for p in points:
        try:
            d = dt.date.fromisoformat(p["date"])
        except (KeyError, ValueError):
            continue
        if p.get("index_pct") is None:
            continue
        out.append({"date": d, "iso": p["date"], "value": p["index_pct"],
                     "backfill": p.get("source") == "daily_backfill_mid"})
    out.sort(key=lambda p: p["date"])
    return out


# A range is only worth its own pill if the window holds enough points to
# show movement, not just its two boundary days. A one-point-per-day series
# can never fill "1h", "6h" or even "1d" this way -- a single daily median
# straddling a one-day window is a dot, not a line -- and that falls out of
# this rule on its own, without hard-coding "this chart is daily".
MIN_POINTS_IN_WINDOW = 4


def available_ranges(points):
    """Range keys the given points can actually fill, "all" always last."""
    pts = _parsed(points)
    if len(pts) < 2:
        return []
    span_days = (pts[-1]["date"] - pts[0]["date"]).days
    out = []
    for key, label, window in RANGES:
        if window is None:
            out.append((key, label))
            continue
        if span_days < window:
            continue
        in_window = [p for p in pts if (pts[-1]["date"] - p["date"]).days <= window]
        if len(in_window) >= MIN_POINTS_IN_WINDOW:
            out.append((key, label))
    return out


def slice_range(points, key):
    pts = _parsed(points)
    if key == "all" or not pts:
        return pts
    window = next((w for k, _, w in RANGES if k == key), None)
    if not window:
        return pts
    cutoff = pts[-1]["date"] - dt.timedelta(days=window)
    return [p for p in pts if p["date"] >= cutoff]


def _runs(pts):
    """Split into contiguous runs: a run ends at a day-gap or a source change."""
    runs = []
    cur = []
    for p in pts:
        if cur and (p["date"] - cur[-1]["date"]).days == 1 and p["backfill"] == cur[-1]["backfill"]:
            cur.append(p)
        else:
            if cur:
                runs.append(cur)
            cur = [p]
    if cur:
        runs.append(cur)
    return runs


def _geometry(pts):
    values = [p["value"] for p in pts]
    lo, hi = min(values + [0]), max(values + [0])
    pad = (hi - lo) * 0.12 or 1
    lo -= pad
    hi += pad
    n = len(pts)

    def x(i):
        return (i / (n - 1)) * WIDTH if n > 1 else 0

    def y(v):
        return HEIGHT - ((v - lo) / (hi - lo)) * HEIGHT

    return lo, hi, x, y


def build(points, *, default_range="all"):
    """Everything the template needs to draw one chart, computed once."""
    ranges = available_ranges(points)
    pts = slice_range(points, default_range if any(k == default_range for k, _ in ranges) else "all")
    if len(pts) < 2:
        return {"has_data": False, "ranges": ranges, "range": default_range}

    lo, hi, x, y = _geometry(pts)
    idx_by_id = {id(p): i for i, p in enumerate(pts)}

    svg_parts = []
    for run in _runs(pts):
        colour = BACKFILL if run[0]["backfill"] else INDIGO
        if len(run) == 1:
            i = idx_by_id[id(run[0])]
            svg_parts.append(
                f'<circle cx="{x(i):.1f}" cy="{y(run[0]["value"]):.1f}" r="2.5" '
                f'fill="{colour}"/>')
            continue
        seg = " L ".join(f"{x(idx_by_id[id(p)]):.1f} {y(p['value']):.1f}" for p in run)
        if not run[0]["backfill"]:
            first_i, last_i = idx_by_id[id(run[0])], idx_by_id[id(run[-1])]
            area = (f'M {x(first_i):.1f} {HEIGHT} L {seg} L {x(last_i):.1f} {HEIGHT} Z')
            svg_parts.append(f'<path d="{area}" fill="url(#chartFade)"/>')
        svg_parts.append(
            f'<path d="M {seg}" fill="none" stroke="{colour}" stroke-width="1.6" '
            f'vector-effect="non-scaling-stroke"/>')

    last = pts[-1]
    badge_y = y(last["value"])
    svg_parts.append(
        f'<line x1="0" x2="{WIDTH}" y1="{badge_y:.1f}" y2="{badge_y:.1f}" '
        f'stroke="{INDIGO}" stroke-width="1" stroke-dasharray="1 3" '
        f'vector-effect="non-scaling-stroke"/>')

    badge_pct = badge_y / HEIGHT * 100
    gutter = []
    for frac, val in ((0.0, hi), (0.5, (hi + lo) / 2), (1.0, lo)):
        top_pct = frac * 100
        if abs(top_pct - badge_pct) < 12:
            continue
        gutter.append({"top_pct": round(top_pct, 1), "text": fmt_pct(val)})

    first = pts[0]
    change_pts = last["value"] - first["value"]
    if round(change_pts, 2) == 0:
        change_text, change_colour = "No change over the period", "var(--muted)"
    else:
        sign = "+" if change_pts >= 0 else "−"
        points_txt = f"{sign}{abs(change_pts):.2f} points"
        # A relative percent is only honest when the value didn't cross zero --
        # past that point "up 400%" is a real division, not a real story.
        if first["value"] != 0 and (first["value"] > 0) == (last["value"] > 0):
            rel = change_pts / abs(first["value"]) * 100
            dir_word = "up" if change_pts >= 0 else "down"
            points_txt += f", {dir_word} {abs(rel):.1f}%"
        change_text = points_txt
        change_colour = INDIGO if change_pts >= 0 else DOWN

    date_idxs = sorted(set(round(f * (len(pts) - 1)) for f in (0, 1 / 3, 2 / 3, 1)))
    dates = []
    for j, i in enumerate(date_idxs):
        align = "start" if j == 0 else ("end" if j == len(date_idxs) - 1 else "middle")
        dates.append({"text": long_date(pts[i]["iso"]), "align": align})

    return {
        "has_data": True,
        "ranges": ranges,
        "range": default_range if any(k == default_range for k, _ in ranges) else "all",
        "headline": fmt_pct(last["value"]),
        "change_text": change_text,
        "change_colour": change_colour,
        "svg": "".join(svg_parts),
        "gutter": gutter,
        "badge": {"top_pct": round(badge_pct, 1), "text": fmt_pct(last["value"])},
        "dates": dates,
        "points_json": [{"date": p["iso"], "index_pct": p["value"],
                          "backfill": p["backfill"]} for p in pts],
    }


def render_html(points, *, root_id="chart"):
    """The full static fragment: fully rendered with no JavaScript at all."""
    c = build(points)
    if not c["has_data"]:
        return (f'<div id="{root_id}" class="chart" data-points="[]">'
                '<p class="muted small">Not enough days yet to draw a line. '
                'Every point collected is shown.</p></div>')

    pills = ""
    if len(c["ranges"]) > 1:
        buttons = "".join(
            f'<button type="button" class="chart-pill{" is-on" if k == c["range"] else ""}" '
            f'data-range="{k}">{label}</button>'
            for k, label in c["ranges"])
        pills = f'<div class="chart-pills">{buttons}</div>'

    gutter = "".join(
        f'<span class="chart-gutter-label" style="top:{g["top_pct"]}%">{g["text"]}</span>'
        for g in c["gutter"])

    dates = "".join(
        f'<span class="chart-date" style="text-align:{d["align"]}">{d["text"]}</span>'
        for d in c["dates"])

    import json as _json
    points_json = _json.dumps(c["points_json"])

    return f"""<div id="{root_id}" class="chart" data-points='{points_json}'>
  <div class="chart-head">
    <div class="chart-value">{c["headline"]}</div>
    <div class="chart-change" style="color:{c["change_colour"]}">{c["change_text"]}</div>
  </div>
  <div class="chart-plot-row">
    <div class="chart-plot-wrap">
      <svg class="chart-svg" viewBox="0 0 {WIDTH} {HEIGHT}" preserveAspectRatio="none">
        <defs>
          <linearGradient id="chartFade" x1="0" y1="0" x2="0" y2="1">
            <stop offset="0%" stop-color="{AREA_TOP}" stop-opacity="0.95"/>
            <stop offset="100%" stop-color="{AREA_TOP}" stop-opacity="0.05"/>
          </linearGradient>
        </defs>
        {c["svg"]}
      </svg>
      <div class="chart-badge" style="top:{c["badge"]["top_pct"]}%">{c["badge"]["text"]}</div>
      <div class="chart-hover" hidden></div>
    </div>
    <div class="chart-gutter">{gutter}</div>
  </div>
  <div class="chart-dates">{dates}</div>
  {pills}
</div>"""
